python/cuml/cuml/internals/outputs.py

```diff
@@ -0,0 +1,439 @@
+        else:
+            res = iu.input_to_cuml_array(res, order="K").array
+
+    if output_type == "cuml":
+        # Return CumlArray/SparseCumlArray directly
+        return res
+
+    if is_sparse:
+        # Coerce output_type to supported sparse types.
+        # Host types -> scipy, cupy otherwise.
+        output_type = "scipy" if output_type in ["numpy", "pandas"] else "cupy"
+
+    return res.to_output(output_type=output_type)
+
+
+def reflect(
+    func=None,
+    *,
+    array=...,
+    model=...,
+    reset=False,
+    skip=False,
+):
+    """Mark a function or method as participating in the reflection system.
+
+    Functions and methods decorated with this get a few additional behaviors:
+
+    - They are run within an "internal API context". This mainly means that
+      reflected functions/methods or estimator fitted attributes will be
+      returned as ``CumlArray`` instances instead of their reflected types. If
+      this is the only behavior you want, decorate with ``reflect(skip=True)``.
+
+    - Their output type is converted to the proper output type following
+      standard cuml behavior. The default behavior covers most cases, but when
+      needed you may want to specify the ``model`` and/or ``array`` parameters
+      manually.
+
+    - For estimators, fit-like methods will store the required metadata like
+      ``_input_type`` to support cases like ``output_type="input"``. To enable
+      this for a method set ``reset=True``.
+
+    Parameters
+    ----------
+    func : callable or None
+        The function to be decorated, or None to curry to be applied later.
+    model : int, str, or None, default=...
+        The ``cuml.Base`` parameter to infer the reflected output type from. By
+        default this will be ``'self'`` (if present), and ``None`` otherwise.
+        Provide a parameter position or name to override. May also provide
+        ``None`` to disable this inference entirely.
+    array : int, str, or None, default=...
+        The array-like parameter to infer the reflected output type from. By
+        default this will be the first argument to the method or function
+        (excluding ``'self'`` or ``model``), or ``None`` if there are no other
+        arguments. Provide a parameter position or name to override. May also
+        provide ``None`` to disable this inference entirely; in this case the
+        output type is expected to be specified manually either internal or
+        external to the method.
+    reset : bool, default=False
+        Set to True for methods like ``fit`` that reset the reflected type on
+        an estimator.
+    skip : bool, default=False
+        Set to True to skip output type inference and processing for a method.
+        This is mostly useful if this step is unnecessary (but the function
+        should still run within an internal API context), or if output type
+        conversion will be handled manually.
```

Comment:
I'm wondering a bit about the naming of this parameter.

To me it does not clearly indicate that it "skips" the output type inference and conversion. Maybe we can make it a bit more explicit, like "`convert_output=False`" ?